Workaround the colorfill problem for the vertical DOS plot

dou-du opened the pull request:
The issue was caused by using "fill:true" for the DOS curves, which gives wired color fill as shown below:

![image](https://user-images.githubusercontent.com/56022756/226248656-b10585bd-9b56-4912-808c-e7aa59926a57.png)

The workaround solution is adding a dumb dataset to represent the y axis. All the DOS curves are filling color to 
the dumb dataset. We put the dumb dataset at the top of the array. So the code is "fill:'-1'". The result is shown below:
 
![image](https://user-images.githubusercontent.com/56022756/226314194-826b742d-fa19-4922-bf9e-d842815b082b.png)

The fill color is set up to the Fermi level.

unkcpz commented:
Thanks @dou-du. Can you describe why the issue happened and why the change will fix the issue? The changes are not self-explained by the code.

unkcpz commented on js/lib/bands.js:
```diff
@@ -170,7 +170,7 @@ function bandPlot(bandDivId, bandPathTextBoxId, dataFilePaths, dosFile, showFerm
                 $("#" + bandDivId + "bt-togglePdos").removeClass("button");
 
-                for (var i = 1; i < theBandPlot.dosSeries.length; i++) {
+                for (var i = 2; i < theBandPlot.dosSeries.length; i++) {
```
What is the index for? Can you add a comment for the local variable here?

dou-du replied:
This is the index to toggle visibility of the PDOS. Since one more dataset is added at the head. We need to start from 2. 1 is the total DOS.

unkcpz replied:
thanks for the explanation, can you add a comment so the developers who take over the maintenance duty in the future can easily understand it?

dou-du replied:
Yes, I add comments in the code.

unkcpz commented on js/lib/bandstructure.js:
```diff
@@ -332,6 +332,12 @@ BandPlot.prototype.initDosChart = function (orientation = 'vertical') {
+                    labels: {
+                        filter: function(item, chart) {
+                        // Logic to remove a particular legend item goes here
+                            return !item.text.includes('test');
```
Is this leftover for test purpose?

dou-du replied:
I will change to a more reasonable label.

unkcpz commented on js/lib/bandstructure.js:
```diff
@@ -804,7 +835,7 @@ BandPlot.prototype.updateDosPlot = function (orientation = 'vertical') {
             borderWidth: 1,
             data: curve,
-            fill: true,
+            fill: '-1',
```
What is this change for?

dou-du replied:
This will fill color between the DOS dataset the dumb dataset (represent y axis).

unkcpz replied:
It is still not very clear why the fill can accept a string with a negative integer number. Can you add a comment above to explain why this change is necessary?

dou-du replied:
Thanks for keeping question about this. I read the instruction again and found a mistake. The "-1" fill color to the relative dataset index. I should use 0, which is the absolute dataset index. 

https://www.chartjs.org/docs/latest/charts/area.html

I have corrected this by using: "fill: 0".

dou-du commented:
> Thanks @dou-du. Can you describe why the issue happened and why the change will fix the issue? The changes are not self-explained by the code.

The exact cause for the issue is unclear. We used "fill: true" before which cases the color filled not corrected. In order to fix it, I create one dumb dataset as x=0 (the y axis). I used the option to fill color between two dataset instead, which means fill color between all DOS datasets and the dumb dataset. This fixed the issue.

unkcpz commented:
> The exact cause for the issue is unclear. We used "fill: true" before which cases the color filled not corrected. In order to fix it, I create one dumb dataset as x=0 (the y axis). I used the option to fill color between two dataset instead, which means fill color between all DOS datasets and the dumb dataset. This fixed the issue.

If the cause of the issue is unclear, I think this is a workaround rather than a fix. But it is fine if you can add clear comments for every change you make. 

Besides, the filled background was for representing the electrons are filled in the energy range, I propose for the filled lines in the range above the fermi energy the background is not filled but using the regular line. Let me know how complicate this is and we see if add this in this PR. To me, the background fill for the whole energy range looks like an improper representation.

unkcpz commented:
Meanwhile, you said this issue might be rooted in the chart.js version, did you check that? what is the status of migration? will updating chart.js solve this issue?

dou-du commented:
> > The exact cause for the issue is unclear. We used "fill: true" before which cases the color filled not corrected. In order to fix it, I create one dumb dataset as x=0 (the y axis). I used the option to fill color between two dataset instead, which means fill color between all DOS datasets and the dumb dataset. This fixed the issue.
> 
> If the cause of the issue is unclear, I think this is a workaround rather than a fix. But it is fine if you can add clear comments for every change you make.
> 
> Besides, the filled background was for representing the electrons are filled in the energy range, I propose for the filled lines in the range above the fermi energy the background is not filled but using the regular line. Let me know how complicate this is and we see if add this in this PR. To me, the background fill for the whole energy range looks like an improper representation.

Sure, I have added description in this PR and comments in the code. I changed the title to "workaround", which is definitely more accurate. :) 

I am going to check how hard is going to fill color to the Fermi level and let you know the result.

dou-du commented:
> Meanwhile, you said this issue might be rooted in the chart.js version, did you check that? what is the status of migration? will updating chart.js solve this issue?

Yes, I had tried. Unfortunately, I even not able to import the Chart.js 4.x to the code correctly. During the Materials Cloud meeting, Giovanni asked me whether want to continue on this task and be clear of my role on this widget with you.

Consider my reminding OSSCAR tasks, I regret to say that I am not able to do the migration of the widget to Chart.js 4.x. 
I also think that is is better to change to plotly.js as you suggested. However, I do not have time to do this either. 
I am going to try my best to clear all the current issues of this version of the widget before my leaving.

dou-du commented:
> > The exact cause for the issue is unclear. We used "fill: true" before which cases the color filled not corrected. In order to fix it, I create one dumb dataset as x=0 (the y axis). I used the option to fill color between two dataset instead, which means fill color between all DOS datasets and the dumb dataset. This fixed the issue.
> 
> If the cause of the issue is unclear, I think this is a workaround rather than a fix. But it is fine if you can add clear comments for every change you make.
> 
> Besides, the filled background was for representing the electrons are filled in the energy range, I propose for the filled lines in the range above the fermi energy the background is not filled but using the regular line. Let me know how complicate this is and we see if add this in this PR. To me, the background fill for the whole energy range looks like an improper representation.

Luckily, fill color up to Fermi level is not hard, here is the result:

![image](https://user-images.githubusercontent.com/56022756/226248182-da850338-158d-4ded-b1a2-3791f34ce38d.png)

unkcpz commented:
@dou-du Thanks a lot!

> Luckily, fill color up to Fermi level is not hard, here is the result:

Very great! This is a very good feature.

> Yes, I had tried. Unfortunately, I even not able to import the Chart.js 4.x to the code correctly. During the Materials Cloud meeting, Giovanni asked me whether want to continue on this task and be clear of my role on this widget with you.

No problem, just let me know is enough and we will think about how to take over your work and continue.

> Consider my reminding OSSCAR tasks, I regret to say that I am not able to do the migration of the widget to Chart.js 4.x.
I also think that is is better to change to plotly.js as you suggested. However, I do not have time to do this either.
I am going to try my best to clear all the current issues of this version of the widget before my leaving.

No problem, I'd open some issue to ask you to add more comment so person who maintain this in future can quickly have his/her hands on.

unkcpz commented on js/lib/bands.js:
```diff
@@ -170,7 +170,8 @@ function bandPlot(bandDivId, bandPathTextBoxId, dataFilePaths, dosFile, showFerm
                 $("#" + bandDivId + "bt-togglePdos").removeClass("button");
 
-                for (var i = 1; i < theBandPlot.dosSeries.length; i++) {
+		// The PDOS curves start from 2. 0 is for the dumb datset of y axis. And 1 is the total DOS
```
```suggestion
		// The PDOS curves start from index 2. 
		// Index 0 is for the dumb datset of y axis add to workaround the step change issue #49 . 
		// Index 1 is for total DOS (The widget don't know if the first set of data is TDOS or not, this need to be polish in future.)
```

unkcpz commented on js/lib/bandstructure.js:
```diff
@@ -784,6 +796,38 @@ BandPlot.prototype.updateDosPlot = function (orientation = 'vertical') {
+    });
+
+    // The color of the dumb dataset is white, label is 'y axis'
+    var dos = {
```
```suggestion
    var dumb_dos = {
```

unkcpz commented on js/lib/bandstructure.js:
```diff
@@ -784,6 +796,38 @@ BandPlot.prototype.updateDosPlot = function (orientation = 'vertical') {
+        label: 'y axis'
+    };
+
+    bandPlotObject.dosSeries.push(dos);
```
```suggestion
    bandPlotObject.dosSeries.push(dumb_dos);
```

unkcpz commented on js/lib/bandstructure.js:
```diff
@@ -784,6 +796,38 @@ BandPlot.prototype.updateDosPlot = function (orientation = 'vertical') {
+    var dosx = bandPlotObject.dosData['dos'][0]['x'];
+    var dosy = bandPlotObject.dosData['dos'][0]['y'];
+
+    dosx.forEach(function (data, k) {
```
Add a comment here to briefly explain why it can fill the background up to fermi energy.
Is the curve.push a special plot function that will draw the line rather than the background?

dou-du replied:
Yes, I add it in the code. Basically, if we only give the dataset up to the Fermi level, the fillcolor will only show for the whole curve range.

unkcpz reviewed (changes requested):
Thanks @dou-du, some minor requests.

unkcpz reviewed (approved):
Thanks! @dou-du
